onnxruntime/contrib_ops/cuda/quantization/qordered_ops/qordered_attention.cc

```diff
@@ -0,0 +1,319 @@
+  // float scale = static_cast<float>(1.0 / const_scale_qkv_layer_[qkv_index]);
+  // CUBLAS_RETURN_IF_ERROR(cublasSscal(CublasHandle(), count, &scale, target, 1));
+  return Status::OK();
+}
+
+Status QOrderedAttention::PrePack(const Tensor& tensor, int input_idx, /*out*/ AllocatorPtr alloc,
+                                  /*out*/ bool& is_packed,
+                                  /*out*/ PrePackedWeights* /*prepacked_weights*/) {
+  is_packed = false;
+
+  if (input_idx == InputIds::ScaleInput) {
+    const_scale_input_ = *tensor.Data<float>();
+  } else if (input_idx >= InputIds::Scale_Q_Gemm && input_idx < InputIds::Scale_Q_Gemm + 3) {
+    const_scale_qkv_layer_[input_idx - InputIds::Scale_Q_Gemm] = *tensor.Data<float>();
+  } else if (input_idx >= InputIds::Q_Weight && input_idx < InputIds::Q_Weight + 3) {
+    is_packed = true;
+    ORT_RETURN_IF_ERROR(PutIntoMergedWeight(tensor, alloc, input_idx - InputIds::Q_Weight));
+  } else if (input_idx >= InputIds::Scale_Q_Weight && input_idx < InputIds::Scale_Q_Weight + 3) {
+    is_packed = true;
+    ORT_RETURN_IF_ERROR(PutIntoMergedWeightScale(tensor, alloc, input_idx - InputIds::Scale_Q_Weight));
+  } else if (input_idx >= InputIds::Q_Bias && input_idx < InputIds::Q_Bias + 3) {
+    is_packed = true;
+    ORT_RETURN_IF_ERROR(PutIntoMergedBias(tensor, alloc, input_idx - InputIds::Q_Bias));
+  } else if (input_idx == InputIds::Scale_QK_Gemm) {
+    float scale = *tensor.Data<float>();
+    double base = std::exp((double)scale / sqrt(qkv_hidden_sizes_[0] / static_cast<double>(num_heads_)));
+    auto* softmax_lookup_data = alloc->Alloc(256 * sizeof(float));
+    softmax_lookup_ = BufferUniquePtr(softmax_lookup_data, BufferDeleter(alloc));
+    ORT_RETURN_IF_ERROR(BuildTableForSoftmaxPowerOf(Stream(), base, (float*)softmax_lookup_.get()));
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+inline void debug_print([[maybe_unused]] const T* arr,
+                        [[maybe_unused]] const size_t sz,
+                        [[maybe_unused]] const int w,
+                        [[maybe_unused]] const char* name) {
+#if defined(DEBUGPRINT_QORDERED_ATTENTION)
+
+  cudaDeviceSynchronize();
+  std::vector<T> buf(sz);
+  cudaMemcpy(buf.data(), arr, sz * sizeof(T), cudaMemcpyDeviceToHost);
+
+  std::cout << "========" << name << std::endl;
+  for (size_t i = 0; i < sz; i++) {
+    if (i % w == 0) std::cout << std::endl;
+    if (std::is_same<T, int8_t>().value) {
+      std::cout << (int)buf[i] << ", ";
+    } else {
+      std::cout << buf[i] << ", ";
+    }
+  }
+  std::cout << std::endl;
+
+#endif
+
+}
+
+#endif
+
+QOrderedAttention::QOrderedAttention(const OpKernelInfo& info) : CudaKernel(info), AttentionBase(info), input_hidden_size_(0) {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11040
+
+  ORT_ENFORCE(qkv_hidden_sizes_.size() == 3, "qkv_hidden_sizes is needed and must be of shape [3]!");
```

Comment:
ORT_ENFORCE(order_weight_ != CUBLAST_ORDER_ROW) as the kernel does not support the row major weights scenario

Reply:
Fixed